Provide own ansible docker images

andyundso opened the pull request:
This PR is a proposal and is not complete yet. But before investing more time into this project, I wanted to have your opinions:

This PR overhauls the current ansible repository. I think there are a few reasons to do this:

1. Python 2 support ends on January 1st, 2020.
2. We currently use the image from [this repository](https://github.com/William-Yeh/docker-ansible), but is not maintained since a year (not replying to issues, not replying to PRs) and so the images are still based on Python 2.
3. Normally, I personally think it isn't too bad to use something which isn't maintained for a while, but Docuteam will use our Ansible image for testing and deploying as well and I think, that puts us in a certain responsibility to deliver up-to-date software / images.
4. Our own ansible roles are currently not tested and Docuteam has interest in our Ruby / Rails role.

So this proposal is mainly driven by the fact that we could use the images at Docuteam, but they would also be helpful to test our own ansible roles against different distributions. If we want to go with this, there would be a few things to add to this PR:

1. Find a few good test cases to verify that the images are working (e.g. adding a playbook and running that)
2. Semaphore pipeline which builds, tests and pushes the images to Docker hub.
3. Trigger the semaphore pipeline every month to have up-to-date images.

tompluess commented on alpine/Dockerfile:
```diff
@@ -0,0 +1,29 @@
+ONBUILD  RUN  echo "===> Updating TLS certificates..."         && \
+    apk --update add openssl ca-certificates
+
+ONBUILD  WORKDIR  /tmp
```
`/tmp` is used as general linux temporary directory.  I would propose to use explicitly `/work` as work directory.

tompluess commented on bionic/Dockerfile:
```diff
@@ -0,0 +1,39 @@
+# SOURCE: https://github.com/William-Yeh/docker-ansible/blob/master/ubuntu18.04-onbuild/Dockerfile
+# There are no offical python ubuntu images available
+# And the PPA of Ansible still relies on Python 2
+FROM ubuntu:18.04
```
`bionic` seems to be the code name of a specific ubuntu linux version.

I personally did not know the code name. and I always mix up the ones for Debian too. 

therefore I would prefer to use something like `ubuntu1804` as folder and later tag name.

(other opinions welcome..)

tompluess commented on buster/Dockerfile:
```diff
@@ -0,0 +1,36 @@
+# SOURCE: https://github.com/William-Yeh/docker-ansible/blob/master/debian9-onbuild/Dockerfile
+FROM python:3.8-slim-buster
```
same here: I would prefer to have `debian9` as identifier, rather than the code name.

and what does `slim` mean? can we omit it?

btw. it seems that the source your referring also has `debian9` in the path.

tompluess reviewed (commented):
Thank you @andyundso, I really appreciate your initiative to overhaul this repository. 

In my opinion, the reasoning and proposals in the description are all pointing to the right direction.

I have some comments regarding naming.

In order to merge this PR, the following steps are required from my perspective:
- [x] add at least one test per image to be built
- [x] add semaphore pipelines, that push automatically from master to docker hub
- [x] also push the alpine image with the `latest` tag
- [x] remove auto builds on https://hub.docker.com/repository/docker/simplificator/ansible/builds

andyundso committed:
Delete existing Dockerfiles

andyundso committed:
Added Dockerfile for building alpine3 image

andyundso committed:
Added Dockerfile for building Debian 10 image

andyundso committed:
Added Dockerfile for building Debian 8 image

andyundso committed:
Added Dockerfile for building Debian 9 image

andyundso committed:
Added Dockerfile for building Ubuntu 18.04 image

andyundso committed:
Build images on Semaphore and push them to Docker Hub

andyundso committed:
Added tests for basic ansible functions

andyundso committed:
Run a test playbook against the Debian based containers

andyundso committed:
Extended README

andyundso committed:
Add SSH to all images

andyundso committed:
Add Gnupg to Debian-based builds

Gnupg is needed for adding GPG keys of repositories to a container. I also made a second script with the special tests for the Debian builds so we can verify that the gpg command is available.

andyundso committed:
Fix version of Ansible in container

andyundso commented:
I just force pushed the whole PR to structure it a bit better, but I tackled all your points @tompluess 

* I added a few tests for the images to check that all required commands are available and are working.
* Semaphore builds the image, tests it and pushes it if the tests pass.
* Auto-builds on Docker Hub are removed.

I also tested the alpine image on our provisioning pipeline to make sure that works.

tompluess reviewed (approved):
Thank you @andyundso - this is great work!
